Updated Arel::Predications, added case_sensitive

naxels opened the pull request:
to does_not_match_any and does_not_match_all

### Summary

I had a need for Arel::Predications does_not_match_any and does_not_match_all functions today
and noticed those were the only matching method's that did not have the case_sensitive argument.

Added the argument and created a sample project to test the output, worked like a charm.

### Other Information

No other information, thank you for continuous improvement to Rails!

Commit:
Updated Arel::Predications, added case_sensitive

to does_not_match_any and does_not_match_all

simi commented:
@naxels I think this will need at least some test to ensure it is not broken in the future. Anyway AREL is currently considered private API and I'm not sure enhancing it with features not used by public API is welcomed.

simi commented:
ℹ️ I just noticed those methods are about to be removed in https://github.com/rails/rails/pull/39150.

naxels commented:
Thank you @simi , I read the other proposal and find it very unfortunate that there is no easy public API way to achieve LIKE and NOT LIKE as easy as it is with the private API, as I most often use these features with Rails.

The problem I actually ran into with the public API is that I cannot achieve the same functionality when combining multiple models that have some overlapping fields, as in my case PostgreSQL will report an error as it doesn't know which table to use.

I had extracted several of these NOT LIKE/LIKE queries into a separate scope file to promote reuse across models, after which it became clear that when joining tables, the error would occur.

I basically translated this (not model independent due to String "ordertype" etc)
      `scope :not_test, -> { where("ordertype NOT LIKE '%Test%'") }`

into this reusable, SQL injection friendly code that now works with joins and can be used in multiple models:
      `scope :not_test, -> { ordertype_not_like('%Test%') }`
      `scope :ordertype_not_like, ->(filter) { field_not_like(:ordertype, filter) }`
      `scope :field_not_like, ->(field, filter) { where(arel_table[field].does_not_match(filter, nil, true)) }`

After making this work, I then had other NOT LIKE and LIKE's that looked like:
      `scope :not_test_multi, -> { where("ordertype NOT LIKE '%Test1%' AND ordertype NOT LIKE 'Test2-%'") }`

Which could now easily translate into (except that it uses ILIKE due to missing case_sensitive), again model independent:
     `scope : not_test_multi, -> { where(arel_table[:ordertype].does_not_match_all(['%Test1%', 'Test2-%'])) }`

Previously when working with LIKE and JOIN I would specify the `table.ordertype` String, which doesn't work with the scope abstraction I made for multiple models.

The reason I added case_sensitive to the 2 Arel methods is due to PostgreSQL Performance of LIKE over ILIKE, since I know the data that will come in.

Maybe you know of a better way to achieve the same?

I would be happy to add the tests if you decide to allow this change to continue.

naxels commented:
Any update on this?
